103_agent-driven-attestation.md

```diff
@@ -0,0 +1,421 @@
+
+Switching between push and pull mode is a separate matter. First, all components which are being switched from one mode to the other must be running a version which supports the new push mode. Then, the user will need to perform the steps below, depending on the component in question:
+
+#### Registrar
+
+The registrar will simultaneously support agents operating in push and pull mode, so it is not necessary to change the mode of the registrar. However, the user may need to review the settings which affect the trust decisions made by the registrar (e.g., by configuring a webhook URI, if that feature is desired).
+
+#### Verifier
+
+To switch a verifier's operating mode, change the appropriate configuration option and restart the verifier. After restarting, all agents which are enrolled with the verifier will be deemed to have failed verification. To restart attestation of the agents, the user will need to change the operating mode of each agent (according the directions below) and then issue a command via the tenant. This is so that the tenant can re-check the trust status of the node's identities before restarting attestation.
+
+#### Agent
+
+To switch an agent's operating mode, again, simply change the agent's configuration and restart the agent. After restarting, the agent will attempt to update its registration at the registrar. If this succeeds and the agent's new operating mode is push, the agent will start sending attestations to the verifier specified in the agent's configuration file. Attestations will be rejected by the verifier until attestations are restarted by issuing the tenant command described above.
+
+### Dependency Requirements
+
+This enhancement proposal requires a relatively significant number of new API endpoints, especially for the verifier. These include those needed to support the new attestation protocol and challenge–response authentication mechanism. Changes to existing endpoints may also be appropriate to expose any new information stored in the database, such as more granular information about the trust status of the individual identities of a node.
+
+It suggested that, just as the Rust agent uses [Actix](https://actix.rs/) to serve its REST API, a similar web micro-framework is added as a dependency to the Python codebase. This will provide a natural way to define routes and request handlers and help newcomers understand how the code is organised.
+
+[Flask](https://github.com/pallets/flask) 2.x would be a good candidate for this, as it is packaged for [Fedora since version 37 and EPEL 9](https://packages.fedoraproject.org/pkgs/python-flask/python3-flask/), [Debian Stable](https://packages.debian.org/bookworm/python3-flask), [Ubuntu since version 22.04](https://packages.ubuntu.com/jammy/python3-flask) and [SLES since version 15 SP1](https://software.opensuse.org/package/python-Flask).
```

Comment:
should not be something more async oriented, like quart (flask rewrite) or trio?

Reply:
I'd have to check the OS support of those, but otherwise I am happy to use any framework. My understanding though is that Tornado is async as well, so we are fine to stick with that, I take it? (see my reply to your previous question)

Reply:
The biggest downside of sticking with Tornado is that its router implementation is very bare bones. Quart allows you to specify routers using string templates, which is rather nice and makes the code super easy to read. But not sure if the community would be amenable to switching frameworks

Reply:
I'm fine with switching frameworks if they is well supported and makes our lives easier.

Reply:
So, I had a look and quart is not currently packaged by Fedora/EPEL unfortunately (it is in Debian and Ubuntu). That's a shame, because quart's expressiveness would have been nice. Trio seems even more bare bones (its not a web framework), so it probably makes the most sense to stick with Tornado (unless there is some other Python web framework with good async support that I don't know about)

Reply:
There is FastAPI, but it has similar packaging issues to quart.  Yeah sticking with Tornado and cleaning up the common code (parameter extraction, message parsing, responses etc.) is probably the best option for now. Also maybe putting the DB interactions in one single place might make sense to make it more robust.

Reply:
Those are all good ideas, thanks Thore. We'll take that approach